File: solutions/python/collatz-conjecture/2/collatz_conjecture.py
def steps(number):
    if number <= 0:
        raise ValueError("Only positive integers are allowed")
    step = 0
    while number != 1:
        match (number & -number).bit_length() - 1:
            case 0:
                number += (number << 1) + 1
                step += 1
            case trailing_zero:
                number >>= trailing_zero
                step += trailing_zero
    return step

File: solutions/lua/collatz-conjecture/2/collatz-conjecture.lua
return function(n)
  assert(n > 0, 'Only positive numbers are allowed')
  local acc = 0
  while n ~= 1 do
    while n & 1 == 0 do
      acc = acc + 1
      n = n >> 1
    end
    if n ~= 1 then
      acc = acc + 1
      n = n + 1 + (n << 1)
    end
  end
  return acc
end